sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeDestructor.scala

```diff
@@ -0,0 +1,259 @@
+
+import scala.collection.immutable.ArraySeq
+
+sealed abstract class TypeDestructor {
+
+  import TypeDestructor._
+
+  type Type <: Ast.Type
+
+  def wrap(typ: Ast.Type): Type
+
+  def destruct(state: Type): Either[Error, TypeF[Type]]
+
+}
+
+object TypeDestructor {
+
+  sealed abstract class TypeF[+Type] extends Product with Serializable
+
+  object TypeF {
+
+    case object UnitF extends TypeF[Nothing]
+
+    case object BoolF extends TypeF[Nothing]
+
+    case object Int64F extends TypeF[Nothing]
+
+    case object DateF extends TypeF[Nothing]
+
+    case object TimestampF extends TypeF[Nothing]
+
+    final case class NumericF(scale: Numeric.Scale) extends TypeF[Nothing]
+
+    case object PartyF extends TypeF[Nothing]
+
+    case object TextF extends TypeF[Nothing]
+
+    final case class ContractIdF[Type](a: Type) extends TypeF[Type]
+
+    final case class OptionalF[Type](a: Type) extends TypeF[Type]
+
+    final case class ListF[Type](a: Type) extends TypeF[Type]
+
+    final case class MapF[Type](a: Type, b: Type) extends TypeF[Type]
+
+    final case class TextMapF[Type](a: Type) extends TypeF[Type]
+
+    final case class RecordF[Type](
+        tyCon: Ref.TypeConName,
+        pkgName: Ref.PackageName,
+        fieldNames: ArraySeq[Ref.Name],
+        fieldTypes: ArraySeq[Type],
+    ) extends TypeF[Type]
+
+    final case class VariantF[Type](
+        tyCon: Ref.TypeConName,
+        pkgName: Ref.PackageName,
+        cons: ArraySeq[Ref.Name],
+        consTypes: ArraySeq[Type],
+    ) extends TypeF[Type] {
+      private[this] lazy val consRankMap = cons.view.zipWithIndex.toMap
+
+      def consRank(cons: Ref.Name): Either[LookupError, Int] = {
+        def ref = Reference.DataEnumConstructor(tyCon, cons)
+
+        consRankMap.get(cons).toRight(LookupError.NotFound(ref, ref))
+      }
+    }
+
+    final case class EnumF(
+        tyCon: Ref.TypeConName,
+        pkgName: Ref.PackageName,
+        cons: ArraySeq[Ref.Name],
+    ) extends TypeF[Nothing] {
+      private[this] lazy val consRankMap = cons.view.zipWithIndex.toMap
+
+      def consRank(cons: Ref.Name): Either[LookupError, Int] = {
+        def ref = Reference.DataEnumConstructor(tyCon, cons)
+
+        consRankMap.get(cons).toRight(LookupError.NotFound(ref, ref))
+      }
+    }
+  }
+
+  sealed abstract class Error extends Product with Serializable
+
+  object Error {
+    final case class LookupError(error: language.LookupError) extends Error
+
+    final case class TypeError(msg: String) extends Error
+  }
+
+  def apply(pkgInterface: PackageInterface): TypeDestructor =
+    new TypeDestructorImpl(pkgInterface)
+
+}
+
+private final class TypeDestructorImpl(pkgInterface: PackageInterface) extends TypeDestructor {
+  self =>
+
+  import TypeDestructor.TypeF
+  import TypeF._
+
+  override type Type = Ast.Type
+
+  override def wrap(typ: Ast.Type): Type = typ
+
+  private var cache: Map[Ast.Type, TypeF[Ast.Type]] = Map.empty
```

Comment:
Is the cache really needed? Is any of the operations costly here? I guess the substitution is a little bit costly, but in practice I don't expect gigantic types.

I'm worried it might introduce memory leaks depending on how it's used.